feat: update labels when config changes (#25340)
renovatebot/renovate: docs/usage/configuration-options.md

With the above config, every PR raised by Renovate will have the label `dependencies` while PRs containing `eslint`-related packages will instead have the label `linting`.

Behaviour details:

> - Renovate only adds labels when it creates the PR

Reviewer: ```suggestion
- Renovate creates PRs with the currently configured/applicable labels
```

> - If you remove labels which Renovate added, it won't re-apply them

Reviewer: ```suggestion
- If you add or remove labels in a PR which Renovate created, it won't touch labels again in that open PR
```

> - If you change your config, the new/changed labels are applied to any open PRs

Reviewer 2: When you say "change your config", do you mean:

- Change label name in Git hosting  platform
- Change label config option in Renovate config file
- Something else

Author: Change label config option in Renovate config file

Reviewer 2: ```suggestion
- If you change the `label` config option, the new/changed labels are applied to any open PRs
```

Author: I think the _before_ is better, because we can also change labels using `addLabels` and within `packageRules`.

With the above config, every PR raised by Renovate will have the label `dependencies` while PRs containing `eslint`-related packages will instead have the label `linting`.

Behaviour details:

> - Renovate creates PRs with the currently configured/applicable labels
> - If you add or remove labels in a PR which Renovate created, it won't touch labels again in that open PR
> - If you change your config, the new/changed labels are applied to any open PRs
> 
> The `labels` array is non-mergeable, meaning if multiple `packageRules` match then Renovate uses the last value for `labels`.
> If you want to add/combine labels, use the `addLabels` config option, which is mergeable.
> 
> <!-- prettier-ignore -->
> !!! note
>     Currently updating of the labels on open PRs is only supported for the platforms: github, gitlab and gitea
> 

Reviewer: ```suggestion
- Renovate creates PRs with labels, if they are configured
- If any other bot or users modifies labels in a PR which Renovate created, it won't touch labels again in that open PR
- If the labels for a PR change through configuration or other means, they are applied to any open PRs with unmodified labels (GitHub, GitLab, and Gitea only)

The `labels` array is non-mergeable, meaning if multiple `packageRules` match then Renovate uses the last value for `labels`.
If you want to add/combine labels, use the `addLabels` config option, which is mergeable.

```

With the above config, every PR raised by Renovate will have the label `dependencies` while PRs containing `eslint`-related packages will instead have the label `linting`.

Behaviour details:

> - Renovate adds labels to the PR at time of PR creation, if any have been configured

Reviewer 2: ```suggestion
- When Renovate creates a PR, it will add any configured labels to that PR
```

> - When another bot, or user, changes the labels on a open Renovate PR then Renovate won't change those labels

Reviewer: ```suggestion
- If another bot, or user, changes the labels on an open Renovate PR then Renovate won't change labels on the PR after that
```

> - If any other bot or users modifies labels in a PR which Renovate created, it won't touch labels again in that open PR

Reviewer 2: ```suggestion
- When another bot, or user, changes the labels on a open Renovate PR then Renovate won't change those labels
```

> - On GitHub, GitLab and Gitea: if the labels for a PR change through configuration , the changes will be applied to all PRs with unmodified labels

Reviewer 2: ```suggestion
- On GitHub, GitLab and Gitea: if the labels for a PR change through configuration, the changes will be applied to all PRs with unmodified labels
```

Fixing a whitespace error.

> - If the labels for a PR change through configuration or other means, they are applied to any open PRs with unmodified labels (GitHub, GitLab, and Gitea only)

Author: ```suggestion
- On GitHub, GitLab and Gitea: if the labels for a PR change through configuration , the changes will be applied to all PRs with unmodified labels
```

> - If the labels for a PR change through configuration or other means, they are applied to any open PRs with unmodified labels (GitHub, GitLab, and Gitea only)

Reviewer 2: ```suggestion
- On GitHub, GitLab and Gitea: if the labels for a PR change through configuration or other means, the changes will be applied to any unmodified labels
```

It's not clear to me what is meant precisely by: `through configuration or other means`. Do you mean something like this?

> If you as user, change the labels in the Renovate config, _or_ via GitHub's, GitLab's or Gitea's UI, then those changes will be applied to any open PRs with unmodified labels?

Author: I only mean through configuration :)

> - On GitHub, GitLab and Gitea: if a PR's labels weren't modified after PR creation, and there is a change to Renovate `labels` configuration for such a PR, then Renovate _will_ reset/reapply configured `labels` to align them with the new config

Reviewer 2: This text is hard to understand. The problem is that we have multiple paths through the code depending on:

- if the user/something else changes the label
- when the label change happens
- on which platforms the label change happens

I think a picture or mermaid graph of the different "routes" would be great! We still need good text for those who can't see the picture or graph though.

Author: It seems okay to me. Not sure a mermaid graph would be needed for this as the logic is similar to how we deal with modified branches. Maybe we could reduce the amount of words here.

Something like:

```md
- Renovate automatically applies configured labels to PRs upon creation.
- If labels on an open Renovate PR are changed manually or through automation after creation, Renovate doesn't enforce its configured labels in subsequent runs.
- If labels remain unchanged after PR creation and Renovate's label configuration is modified, Renovate will update the labels to match the new configuration on GitHub, GitLab, and Gitea.
```

> - Renovate will add any configured labels to PRs at time of PR creation
> - If labels on an open Renovate PR are modified after PR creation (e.g. via a user manually, or any automated process) then Renovate won't "enforce" its configured labels on subsequent runs. i.e. it won't reset/reapply its configured `labels`
> - On GitHub, GitLab and Gitea: if a PR's labels weren't modified after PR creation, and there is a change to Renovate `labels` configuration for such a PR, then Renovate _will_ reset/reapply configured `labels` to align them with the new config

Reviewer: ```suggestion
- On GitHub, GitLab and Gitea: Renovate will keep PR labels in sync with configured labels, provided that no other user or bot has made changes to the labels after PR creation. If labels are changed by any other account, Renovate will stop making further changes.
- For other platforms, Renovate will add labels only at time of PR creation and not update them after that.
```
